modules/dxp/apps/segments/segments-asah-connector/src/main/java/com/liferay/segments/asah/connector/internal/service/SegmentsEntryLocalServiceWrapper.java

```diff
@@ -0,0 +1,163 @@
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.segments.asah.connector.internal.client.AsahFaroBackendClient;
+import com.liferay.segments.asah.connector.internal.client.AsahFaroBackendClientImpl;
+import com.liferay.segments.asah.connector.internal.client.model.IndividualSegment;
+import com.liferay.segments.asah.connector.internal.expression.IndividualSegmentsExpressionVisitorImpl;
+import com.liferay.segments.asah.connector.internal.expression.parser.IndividualSegmentsExpressionLexer;
+import com.liferay.segments.asah.connector.internal.expression.parser.IndividualSegmentsExpressionParser;
+import com.liferay.segments.constants.SegmentsEntryConstants;
+import com.liferay.segments.criteria.Criteria;
+import com.liferay.segments.criteria.CriteriaSerializer;
+import com.liferay.segments.model.SegmentsEntry;
+
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+
+import org.antlr.v4.runtime.ANTLRInputStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Mikel Lorza
+ */
+@Component(service = ServiceWrapper.class)
+public class SegmentsEntryLocalServiceWrapper
+	extends com.liferay.segments.service.SegmentsEntryLocalServiceWrapper {
+
+	@Override
+	public SegmentsEntry recalculateSegmentsEntry(long segmentsEntryId)
+		throws PortalException {
+
+		SegmentsEntry segmentsEntry = super.recalculateSegmentsEntry(
+			segmentsEntryId);
+
+		try {
+			if (!FeatureFlagManagerUtil.isEnabled("LPS-172194") ||
+				!_analyticsSettingsManager.isAnalyticsEnabled(
+					segmentsEntry.getCompanyId()) ||
+				!SegmentsEntryConstants.SOURCE_ASAH_FARO_BACKEND.equals(
+					segmentsEntry.getSource())) {
+
+				return segmentsEntry;
+			}
+		}
+		catch (Exception exception) {
+			_log.error(exception);
+
+			return segmentsEntry;
+		}
+
+		IndividualSegment individualSegment =
+			_asahFaroBackendClient.getIndividualSegment(
+				segmentsEntry.getCompanyId(),
+				segmentsEntry.getSegmentsEntryKey());
+
+		IndividualSegmentsExpressionParser individualSegmentsExpressionParser =
+			new IndividualSegmentsExpressionParser(
+				new CommonTokenStream(
+					new IndividualSegmentsExpressionLexer(
+						new ANTLRInputStream(individualSegment.getFilter()))));
+
+		IndividualSegmentsExpressionParser.ExpressionContext expressionContext =
+			individualSegmentsExpressionParser.expression();
+
+		Criteria criteria = expressionContext.accept(
+			new IndividualSegmentsExpressionVisitorImpl());
+
+		ServiceContext serviceContext = _getServiceContext(
+			segmentsEntry.getCompanyId());
+
+		Map<Locale, String> nameMap = Collections.singletonMap(
+			_portal.getSiteDefaultLocale(serviceContext.getScopeGroupId()),
+			individualSegment.getName());
+
+		return updateSegmentsEntry(
```

Comment:
@Mikellorza, @cgoncas, one question, can we call the service here in case there's a wrapper for updateSegmentsEntry too? Like: _segmentsEntryLocalService.updateSegmentsEntry?

Reply:
Don't pay attention to my comment, this would only happen in the serviceImpl, not in the wrapper

Reply:
Hey @jkappler !!

I think this is a very interesting question, the reason I have decided to call directly updateSegmentsEntry instead _segmentsEntryLocalService.updateSegmentsEntry is that:

1. _SegmentsEntryLocalServiceWrapper_ extends _com.liferay.segments.service.SegmentsEntryLocalServiceWrapper_, so if you call **updateSegmentsEntry** you are calling the one implemented in the current _SegmentsEntryLocalServiceWrapper_ or if you don't have any implementation of updateSegmentsEntry in this class, the one in _com.liferay.segments.service.SegmentsEntryLocalServiceWrapper_ will be called.
1. If we have another SegmentsEntryLocalServiceWrapper that doesn't extend the current _SegmentsEntryLocalServiceWrapper_, I don't know how the AopInvocationHandler resolves that case exactly. 

Thanks a lot for the question.

Reply:
great, thanks @cgoncas!